src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_net_command_factory_json.py

```diff
@@ -1,13 +1,68 @@
+import os
+
 from _pydev_bundle._pydev_imports_tipper import TYPE_IMPORT, TYPE_CLASS, TYPE_FUNCTION, TYPE_ATTR, \
     TYPE_BUILTIN, TYPE_PARAM
 from _pydev_bundle.pydev_is_thread_alive import is_thread_alive
 from _pydev_bundle.pydev_override import overrides
 from _pydevd_bundle._debug_adapter import pydevd_schema
-from _pydevd_bundle.pydevd_comm_constants import CMD_THREAD_CREATE, CMD_RETURN
-from _pydevd_bundle.pydevd_constants import get_thread_id
+from _pydevd_bundle.pydevd_comm_constants import CMD_THREAD_CREATE, CMD_RETURN, CMD_MODULE_EVENT
+from _pydevd_bundle.pydevd_constants import get_thread_id, dict_values
 from _pydevd_bundle.pydevd_net_command import NetCommand
 from _pydevd_bundle.pydevd_net_command_factory_xml import NetCommandFactory
 from _pydevd_bundle.pydevd_utils import get_non_pydevd_threads
+from _pydev_imps._pydev_saved_modules import threading
+from _pydevd_bundle._debug_adapter.pydevd_schema import ModuleEvent, ModuleEventBody, Module
+from functools import partial
+import itertools
+import pydevd_file_utils
+
+
+class ModulesManager(object):
+
+    def __init__(self):
+        self._lock = threading.Lock()
+        self._modules = {}
+        self._next_id = partial(next, itertools.count(0))
+
+    def track_module(self, filename_in_utf8, module_name, frame):
+        '''
+        :return list(NetCommand):
+            Returns a list with the module events to be sent.
+        '''
+        if filename_in_utf8 in self._modules:
+            return []
+
+        module_events = []
+        with self._lock:
+            # Must check again after getting the lock.
+            if filename_in_utf8 in self._modules:
+                return
+
+            version = frame.f_globals.get('__version__', '')
+            package_name = frame.f_globals.get('__package__', '')
+            module_id = self._next_id()
+
+            module = Module(module_id, module_name, filename_in_utf8)
+            if version:
+                module.version = version
+
+            if package_name:
+                # Note: package doesn't appear in the docs but seems to be expected?
+                module.kwargs['package'] = package_name
+
+            module_event = ModuleEvent(ModuleEventBody('new', module))
```

Comment:
Since path to module is known here, you might want to add `path` to the module event body as well.

Reply:
I've created https://github.com/Microsoft/ptvsd/issues/1172 to address this (trying to keep the features in this pull request as close as possible to what was already available).

Reply:
Actually, `filename_in_utf8` is passed as the path already, so, closing #1172.